### PR TITLE
✨  add health check for webhook server

### LINKED ISSUE #723: Add default readiness check for webhook server?
In the Cluster API project, we're trying to automate the following in our CI testing:

1. Create a kind cluster
1. Deploy cert-manager
1. Wait for cert-manager to be ready
1. Deploy Cluster API (CRDs, RBAC, Deployments, etc)
1. Wait for Cluster API pods to be ready (so our webhooks are ready)
1. Create a `Cluster`, `Machine`, etc.

When we're waiting for the Cluster API pods to be ready, we've noticed some test flakes because it appears the pods are ready but their webhook servers are not fully online and ready to respond to requests yet.

I've thought about adding a readiness probe to our Manager, but I believe there is still a chance that it could return a false positive; i.e., the check returns ready, but because the webhook server is technically a different server than the readiness server, the webhook server might not be ready yet.

Would it make sense to add a default readiness check for the webhook server?

### PR DESCRIPTION
**Pull request by sbueringer**
Signed-off-by: Stefan Büringer buringerst@vmware.com

Fixes #723 

Supersedes #1124

### CODE REVIEW

**Comment by sbueringer** [comment]
I implemented it this way as it's the easiest implementation and there didn't seem to be major concerns in #1124 (which has a similar implementation)

My biggest concern is that the `started` field is not 100% correct as the `Serve` func is executed after setting `started` to true. An alternative would be to check if the server has been started by trying a tcp tial on the webhook port (similar to https://github.com/kubernetes-sigs/cluster-api/blob/master/internal/envtest/environment.go#L225-L241).

File: pkg/webhook/server.go [hunk 1]
```diff
@@ -280,6 +288,19 @@ func (s *Server) Start(ctx context.Context) error {
 	return nil
 }
 
+// StartedChecker returns an healthz.Checker which is healthy after the
+// server has been started.
+func (s *Server) StartedChecker() healthz.Checker {
```
Comment (sbueringer) [inline pkg/webhook/server.go:293]: I can also move this func into the healthz package. The consequence would be that either the healthz package has a dependency onto the webhook package or I would have to add an interface with a `Started() bool` func.

Comment (vincepri) [reply pkg/webhook/server.go:293]: Could we maybe also tcp ping the server?

Comment (sbueringer) [reply pkg/webhook/server.go:293]: Yup definitely. Should we then still keep the started flag? 

Or is the idea to combine that, so we only start pinging after started is true?

I'm not sure if it's worth it keeping the started field around.

Comment (vincepri) [reply pkg/webhook/server.go:293]: It can make sense to only start tcp-pinging the port only after the started field is set to true

Comment (sbueringer) [reply pkg/webhook/server.go:293]: Yup, will do.

**Comment by sbueringer** [comment]
I'll squash before merge.

File: pkg/webhook/server.go [hunk 2]
```diff
@@ -280,6 +289,27 @@ func (s *Server) Start(ctx context.Context) error {
...
+	return func(req *http.Request) error {
+		s.mu.Lock()
+		defer s.mu.Unlock()
+
+		if !s.started {
+			return fmt.Errorf("webhook server has not been started yet")
+		}
+
+		conn, err := net.DialTimeout("tcp", net.JoinHostPort(s.Host, strconv.Itoa(s.Port)), 10*time.Second)
```
Comment (sbueringer) [inline pkg/webhook/server.go:303]: 10 seconds is just a wild guess. Not sure after which threshold we want to declare the webhook server not ready.

**Review by cndoit18 — APPROVED** [review]

**Review by vincepri — APPROVED** [review]
/approve
/lgtm

**Pull request merged**